File: src-tauri/src/cua.rs
//! Computer Use Agent (CUA) bridge.
//!
//! Implements the GPT-5.5 computer tool loop via the Responses API:
//! 1. Send task + screenshot to GPT-5.5 with `tools: [{ type: "computer" }]`
//! 2. Receive `computer_call` with `actions[]`
//! 3. Execute each action on the Playwright sidecar via `browser_command`
//! 4. Capture updated screenshot → send as `computer_call_output`
//! 5. Repeat until the model stops returning `computer_call`
//!
//! The existing browser.rs Playwright sidecar handles action execution.

use serde::{Deserialize, Serialize};
use std::fs;
use std::process::Command;

/// Maximum loop iterations to prevent runaway sessions
const MAX_TURNS: usize = 30;
/// Viewport dimensions (must match browser-agent.ts)
const VIEWPORT_W: u32 = 1280;
const VIEWPORT_H: u32 = 900;

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct CuaProgress {
    pub turn: usize,
    pub action_count: usize,
    pub status: String,
    pub summary: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct CuaResult {
    pub ok: bool,
    pub turns_used: usize,
    pub summary: String,
    pub final_screenshot_base64: Option<String>,
}

/// Run a full CUA session: model plans + acts via the Playwright browser.
///
/// Called from the frontend. Uses `browser_command` (from browser.rs)
/// to execute actions and capture screenshots.
#[tauri::command]
pub async fn cua_run(task: String, url: Option<String>) -> Result<CuaResult, String> {
    let config = crate::commands::read_config_internal()?;
    let api_key = config.api_key.ok_or("No API key configured")?;

    // Ensure browser is open; navigate if URL given
    if let Some(ref u) = url {
        crate::browser::browser_command("open".into(), serde_json::json!({ "url": u })).await?;
    }

    // Take initial screenshot
    let init_ss = take_screenshot().await?;

    // First Responses API call
    let first_body = serde_json::json!({
        "model": "gpt-5.5",
        "tools": [{
            "type": "computer",
            "computer": {
                "display_width": VIEWPORT_W,
                "display_height": VIEWPORT_H,
                "environment": "browser"
            }
        }],
        "input": [
            {
                "role": "user",
                "content": [
                    { "type": "text", "text": task },
                    {
                        "type": "input_image",
                        "image_url": format!("data:image/png;base64,{}", init_ss.base64),
                        "detail": "original"
                    }
                ]
            }
        ],
        "reasoning": { "effort": "medium" },
        "max_output_tokens": 2048
    });

    let mut resp = call_responses_api(&api_key, &first_body)?;
    let mut prev_resp_id = extract_response_id(&resp);
    let mut turns: usize = 0;
    let mut last_screenshot_b64: Option<String> = Some(init_ss.base64);

    loop {
        turns += 1;
        if turns > MAX_TURNS {
            return Ok(CuaResult {
                ok: true,
                turns_used: turns,
                summary: "Reached maximum turns. Task may be partially complete.".into(),
                final_screenshot_base64: last_screenshot_b64,
            });
        }

        // Look for computer_call in the output
        let computer_call = find_computer_call(&resp);
        if computer_call.is_none() {
            // Model is done — extract its text summary
            let summary = extract_text_output(&resp);
            return Ok(CuaResult {
                ok: true,
                turns_used: turns,
                summary,
                final_screenshot_base64: last_screenshot_b64,
            });
        }

        let cc = computer_call.unwrap();
        let call_id = cc.get("call_id").and_then(|v| v.as_str()).unwrap_or("").to_string();
        let actions = cc.get("actions").and_then(|v| v.as_array()).cloned().unwrap_or_default();

        eprintln!("[cua] turn {turns}: executing {} actions", actions.len());

        // Execute each action via the Playwright sidecar
        for action in &actions {
            execute_cua_action(action).await?;
        }

        // Capture updated screenshot
        let ss = take_screenshot().await?;
        last_screenshot_b64 = Some(ss.base64.clone());

        // Build follow-up request with previous_response_id
        let follow_body = serde_json::json!({
            "model": "gpt-5.5",
            "tools": [{
                "type": "computer",
                "computer": {
                    "display_width": VIEWPORT_W,
                    "display_height": VIEWPORT_H,
                    "environment": "browser"
                }
            }],
            "previous_response_id": prev_resp_id,
            "input": [{
                "type": "computer_call_output",
                "call_id": call_id,
                "output": {
                    "type": "input_image",
                    "image_url": format!("data:image/png;base64,{}", ss.base64),
                    "detail": "original"
                }
            }],
            "reasoning": { "effort": "medium" },
            "max_output_tokens": 2048
        });

        resp = call_responses_api(&api_key, &follow_body)?;
        prev_resp_id = extract_response_id(&resp);
    }
}

// ── Helpers ──────────────────────────────────────────────────────────────

struct Screenshot {
    base64: String,
}

async fn take_screenshot() -> Result<Screenshot, String> {
    let result = crate::browser::browser_command(
        "cua_screenshot".into(),
        serde_json::Value::Object(serde_json::Map::new()),
    )
    .await?;

    if !result.ok {
        return Err("Screenshot failed".into());
    }

    let b64 = result
        .data
        .get("base64")
        .and_then(|v| v.as_str())
        .ok_or("No base64 in screenshot response")?
        .to_string();

    Ok(Screenshot { base64: b64 })
}

/// Execute a single CUA action from the GPT-5.5 computer_call
async fn execute_cua_action(action: &serde_json::Value) -> Result<(), String> {
    let action_type = action
        .get("type")
        .and_then(|v| v.as_str())
        .unwrap_or("unknown");

    match action_type {
        "click" => {
            let x = action.get("x").and_then(|v| v.as_f64()).unwrap_or(0.0);
            let y = action.get("y").and_then(|v| v.as_f64()).unwrap_or(0.0);
            let button = action
                .get("button")
                .and_then(|v| v.as_str())
                .unwrap_or("left");
            let keys = action
                .get("keys")
                .and_then(|v| v.as_array())
                .map(|a| a.iter().filter_map(|k| k.as_str().map(String::from)).collect::<Vec<_>>())
                .unwrap_or_default();

            let params = serde_json::json!({
                "x": x as i32,
                "y": y as i32,
                "button": button,
                "keys": keys,
            });
            crate::browser::browser_command("cua_click".into(), params).await?;
        }
        "double_click" => {
            let x = action.get("x").and_then(|v| v.as_f64()).unwrap_or(0.0);
            let y = action.get("y").and_then(|v| v.as_f64()).unwrap_or(0.0);
            let params = serde_json::json!({ "x": x as i32, "y": y as i32 });
            crate::browser::browser_command("cua_double_click".into(), params).await?;
        }
        "type" => {
            let text = action
                .get("text")
                .and_then(|v| v.as_str())
                .unwrap_or("");
            let params = serde_json::json!({ "text": text });
            crate::browser::browser_command("cua_type".into(), params).await?;
        }
        "keypress" => {
            let keys = action
                .get("keys")
                .and_then(|v| v.as_array())
                .map(|a| a.iter().filter_map(|k| k.as_str().map(String::from)).collect::<Vec<_>>())
                .unwrap_or_default();
            let params = serde_json::json!({ "keys": keys });
            crate::browser::browser_command("cua_keypress".into(), params).await?;
        }
        "scroll" => {
            let x = action.get("x").and_then(|v| v.as_f64()).unwrap_or(640.0);
            let y = action.get("y").and_then(|v| v.as_f64()).unwrap_or(450.0);
            let sx = action
                .get("scroll_x")
                .and_then(|v| v.as_f64())
                .unwrap_or(0.0);
            let sy = action
                .get("scroll_y")
                .and_then(|v| v.as_f64())
                .unwrap_or(0.0);
            let params =
                serde_json::json!({ "x": x as i32, "y": y as i32, "scroll_x": sx as i32, "scroll_y": sy as i32 });
            crate::browser::browser_command("cua_scroll".into(), params).await?;
        }
        "drag" => {
            let path = action
                .get("path")
                .cloned()
                .unwrap_or(serde_json::Value::Array(vec![]));
            let params = serde_json::json!({ "path": path });
            crate::browser::browser_command("cua_drag".into(), params).await?;
        }
        "move" => {
            let x = action.get("x").and_then(|v| v.as_f64()).unwrap_or(0.0);
            let y = action.get("y").and_then(|v| v.as_f64()).unwrap_or(0.0);
            let params = serde_json::json!({ "x": x as i32, "y": y as i32 });
            crate::browser::browser_command("cua_move".into(), params).await?;
        }
        "wait" => {
            let ms = action.get("ms").and_then(|v| v.as_u64()).unwrap_or(2000);
            let params = serde_json::json!({ "ms": ms });
            crate::browser::browser_command("cua_wait".into(), params).await?;
        }
        "screenshot" => {
            // Model just wants a screenshot; we always send one after the batch
        }
        other => {
            eprintln!("[cua] unknown action type: {other}");
        }
    }

    Ok(())
}

/// Call the OpenAI Responses API
fn call_responses_api(
    api_key: &str,
    body: &serde_json::Value,
) -> Result<serde_json::Value, String> {
    let body_str = serde_json::to_string(body).map_err(|e| format!("JSON: {e}"))?;
    let tmp = "/tmp/samuel-cua-request.json";
    fs::write(tmp, &body_str).map_err(|e| format!("Write tmp: {e}"))?;

    let output = Command::new("/usr/bin/curl")
        .args([
            "-s",
            "-X",
            "POST",
            "https://api.openai.com/v1/responses",
            "-H",
            &format!("Authorization: Bearer {api_key}"),
            "-H",
            "Content-Type: application/json",
            "-d",
            &format!("@{tmp}"),
        ])
        .output()
        .map_err(|e| format!("curl: {e}"))?;

    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        let stdout = String::from_utf8_lossy(&output.stdout);
        return Err(format!("CUA API error: {stderr} {stdout}"));
    }

    let resp: serde_json::Value =
        serde_json::from_slice(&output.stdout).map_err(|e| format!("Parse CUA response: {e}"))?;

    if let Some(err) = resp.get("error") {
        return Err(format!("CUA API error: {err}"));
    }

    Ok(resp)
}

fn extract_response_id(resp: &serde_json::Value) -> String {
    resp.get("id")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .to_string()
}

/// Find a `computer_call` item in the response output array
fn find_computer_call(resp: &serde_json::Value) -> Option<serde_json::Value> {
    let output = resp.get("output")?.as_array()?;
    for item in output {
        if item.get("type").and_then(|v| v.as_str()) == Some("computer_call") {
            return Some(item.clone());
        }
    }
    None
}

/// Extract text output from the response (model's final answer)
fn extract_text_output(resp: &serde_json::Value) -> String {
    let output = match resp.get("output").and_then(|v| v.as_array()) {
        Some(arr) => arr,
        None => return "Task completed.".to_string(),
    };

    let mut texts = Vec::new();
    for item in output {
        match item.get("type").and_then(|v| v.as_str()) {
            Some("message") => {
                if let Some(content) = item.get("content").and_then(|v| v.as_array()) {
                    for c in content {
                        if c.get("type").and_then(|v| v.as_str()) == Some("output_text") {
                            if let Some(t) = c.get("text").and_then(|v| v.as_str()) {
                                texts.push(t.to_string());
                            }
                        }
                    }
                }
            }
            _ => {}
        }
    }

    if texts.is_empty() {
        "Task completed.".to_string()
    } else {
        texts.join("\n")
    }
}

File: electron/handlers/cua.ts
import { execFileSync } from "node:child_process";
import { existsSync, readFileSync, unlinkSync } from "node:fs";
import { join, resolve } from "node:path";
import { browser_command } from "./browser.js";
import { getWindowRef } from "../window-ref.js";
import { openaiFetch } from "./openai-client.js";

const MAX_TURNS = 30;
// Resampling width for native-mode screenshots before sending to the model.
// 1280 is wide enough that the model can resolve normal UI text but small
// enough to keep the JPEG payload under a few hundred KB. Coordinate output
// from the model is then scaled back up by `scaleX`/`scaleY` to real pixels.
const NATIVE_W = 1280;

// ── Types ───────────────────────────────────────────────────────────────

interface CuaResult {
	ok: boolean;
	turns_used: number;
	summary: string;
	final_screenshot_base64: string | null;
}

interface ComputerCall {
	call_id: string;
	actions: Record<string, unknown>[];
}

interface DisplayDimensions {
	width: number;
	height: number;
}

// ── Browser-mode CUA ────────────────────────────────────────────────────

export async function cua_run(task: string, url?: string): Promise<CuaResult> {
	if (url) {
		await browser_command("open", { url });
	}

	const initSs = await takeBrowserScreenshot();

	// GA computer-use wire shape: tool selector "computer" with no
	// display_width/display_height/environment — the model reads dimensions
	// from the screenshot itself. The earlier shape (computer_use_preview +
	// display_width + display_height + environment) only works against the
	// computer-use-preview model; mixing those preview-only fields with the
	// GA "computer" selector returns
	//   "Unknown parameter: 'tools[0].display_width'"
	// which throws and bubbles up as a tool-call failure to the user.
	// See https://openai.github.io/openai-agents-python/tools/ for the
	// model→wire-shape matrix.
	const firstBody = {
		model: "gpt-5.5",
		tools: [{ type: "computer" }],
		input: [
			{
				role: "user",
				content: [
					{ type: "input_text", text: task },
					{
						type: "input_image",
						image_url: `data:image/png;base64,${initSs}`,
						detail: "original",
					},
				],
			},
		],
		reasoning: { effort: "medium" },
		max_output_tokens: 2048,
	};

	let resp = await callResponsesApi(firstBody);
	let prevRespId = resp.id ?? "";
	let turns = 0;
	let lastScreenshotB64: string | null = initSs;

	for (;;) {
		turns++;
		if (turns > MAX_TURNS) {
			return {
				ok: true,
				turns_used: turns,
				summary: "Reached maximum turns. Task may be partially complete.",
				final_screenshot_base64: lastScreenshotB64,
			};
		}

		const cc = findComputerCall(resp);
		if (!cc) {
			return {
				ok: true,
				turns_used: turns,
				summary: extractTextOutput(resp),
				final_screenshot_base64: lastScreenshotB64,
			};
		}

		console.error(`[cua] turn ${turns}: executing ${cc.actions.length} actions`);

		for (const action of cc.actions) {
			await executeCuaAction(action);
		}

		const ss = await takeBrowserScreenshot();
		lastScreenshotB64 = ss;

		const followBody = {
			model: "gpt-5.5",
			tools: [{ type: "computer" }],
			previous_response_id: prevRespId,
			input: [
				{
					type: "computer_call_output",
					call_id: cc.call_id,
					output: {
						type: "input_image",
						image_url: `data:image/png;base64,${ss}`,
						detail: "original",
					},
				},
			],
			reasoning: { effort: "medium" },
			max_output_tokens: 2048,
		};

		resp = await callResponsesApi(followBody);
		prevRespId = resp.id ?? "";
	}
}

// ── Native-mode CUA ────────────────────────────────────────────────────

export async function cua_run_native(task: string, app?: string): Promise<CuaResult> {
	const { width: screenW, height: screenH } = getDisplayPointDimensions();
	const screenshotH = Math.round((NATIVE_W * screenH) / screenW);
	const scaleX = screenW / NATIVE_W;
	const scaleY = screenH / screenshotH;
	console.error(
		`[cua-native] screenshot: ${NATIVE_W}x${screenshotH}, screen: ${screenW}x${screenH}, scale: ${scaleX.toFixed(2)}x${scaleY.toFixed(2)}`,
	);

	// Hide the Samuel window so it doesn't appear in screenshots or block clicks
	const win = getWindowRef();
	if (win) {
		win.hide();
		await sleep(200);
	}

	if (app) {
		try {
			execFileSync("/usr/bin/osascript", ["-e", `tell application "${app}" to activate`]);
		} catch {
			// best effort
		}
		await sleep(500);
	}

	try {
		const initSs = native_screenshot();

		const firstBody = {
			model: "gpt-5.5",
			tools: [{ type: "computer" }],
			input: [
				{
					role: "user",
					content: [
						{ type: "input_text", text: task },
						{
							type: "input_image",
							image_url: `data:image/jpeg;base64,${initSs}`,
							detail: "original",
						},
					],
				},
			],
			reasoning: { effort: "medium" },
			max_output_tokens: 2048,
		};

		let resp = await callResponsesApi(firstBody);
		let prevRespId = resp.id ?? "";
		let turns = 0;
		let lastScreenshotB64: string | null = initSs;

		for (;;) {
			turns++;
			if (turns > MAX_TURNS) {
				return {
					ok: true,
					turns_used: turns,
					summary: "Reached maximum turns. Task may be partially complete.",
					final_screenshot_base64: lastScreenshotB64,
				};
			}

			const cc = findComputerCall(resp);
			if (!cc) {
				return {
					ok: true,
					turns_used: turns,
					summary: extractTextOutput(resp),
					final_screenshot_base64: lastScreenshotB64,
				};
			}

			console.error(`[cua-native] turn ${turns}: executing ${cc.actions.length} actions`);

			for (const action of cc.actions) {
				native_computer_action(action, scaleX, scaleY);
			}

			await sleep(300);

			const ss = native_screenshot();
			lastScreenshotB64 = ss;

			const followBody = {
				model: "gpt-5.5",
				tools: [{ type: "computer" }],
				previous_response_id: prevRespId,
				input: [
					{
						type: "computer_call_output",
						call_id: cc.call_id,
						output: {
							type: "input_image",
							image_url: `data:image/jpeg;base64,${ss}`,
							detail: "original",
						},
					},
				],
				reasoning: { effort: "medium" },
				max_output_tokens: 2048,
			};

			resp = await callResponsesApi(followBody);
			prevRespId = resp.id ?? "";
		}
	} finally {
		// Always restore the Samuel window after CUA completes
		if (win) {
			win.show();
		}
	}
}

// ── Native helpers (exported for direct use) ────────────────────────────

export function native_screenshot(): string {
	const tmpPng = "/tmp/samuel-cua-native-ss.png";
	const tmpJpg = "/tmp/samuel-cua-native-ss.jpg";
	try {
		unlinkSync(tmpPng);
	} catch {
		// ignore
	}

	const displayIdx = getDefaultDisplay();
	execFileSync("/usr/sbin/screencapture", ["-x", `-D${displayIdx}`, tmpPng]);

	try {
		execFileSync("/usr/bin/sips", [
			"--resampleWidth",
			String(NATIVE_W),
			"--setProperty",
			"format",
			"jpeg",
			"--setProperty",
			"formatOptions",
			"50",
			tmpPng,
			"--out",
			tmpJpg,
		]);
	} catch {
		throw new Error("Failed to resize native screenshot");
	} finally {
		try {
			unlinkSync(tmpPng);
		} catch {
			// ignore
		}
	}

	const data = readFileSync(tmpJpg);
	try {
		unlinkSync(tmpJpg);
	} catch {
		// ignore
	}
	console.error(`[cua-native] screenshot: ${data.length} bytes`);
	return data.toString("base64");
}

export function native_computer_action(
	action: Record<string, unknown>,
	scaleX: number,
	scaleY: number,
): void {
	const actionType = (action.type as string) ?? "unknown";
	const helper = findNativeInputHelper();

	switch (actionType) {
		case "click": {
			const x = ((action.x as number) ?? 0) * scaleX;
			const y = ((action.y as number) ?? 0) * scaleY;
			const button = (action.button as string) ?? "left";
			console.error(`[cua-native] click (${x.toFixed(0)}, ${y.toFixed(0)}) [${button}]`);
			runNativeInput(helper, ["click", String(x), String(y), button]);
			break;
		}
		case "double_click": {
			const x = ((action.x as number) ?? 0) * scaleX;
			const y = ((action.y as number) ?? 0) * scaleY;
			console.error(`[cua-native] double_click (${x.toFixed(0)}, ${y.toFixed(0)})`);
			runNativeInput(helper, ["double_click", String(x), String(y)]);
			break;
		}
		case "type": {
			const text = (action.text as string) ?? "";
			runNativeInput(helper, ["type", text]);
			break;
		}
		case "keypress": {
			const keys = (action.keys as string[]) ?? [];
			runNativeInput(helper, ["keypress", ...keys]);
			break;
		}
		case "scroll": {
			const x = ((action.x as number) ?? 640) * scaleX;
			const y = ((action.y as number) ?? 450) * scaleY;
			const sx = (action.scroll_x as number) ?? 0;
			const sy = (action.scroll_y as number) ?? 0;
			runNativeInput(helper, ["scroll", String(x), String(y), String(sx), String(sy)]);
			break;
		}
		case "drag": {
			const path = action.path as number[][] | undefined;
			if (path) {
				const args = ["drag"];
				for (const pt of path) {
					if (pt.length >= 2) {
						args.push(String((pt[0] ?? 0) * scaleX));
						args.push(String((pt[1] ?? 0) * scaleY));
					}
				}
				runNativeInput(helper, args);
			}
			break;
		}
		case "move": {
			const x = ((action.x as number) ?? 0) * scaleX;
			const y = ((action.y as number) ?? 0) * scaleY;
			runNativeInput(helper, ["move", String(x), String(y)]);
			break;
		}
		case "wait": {
			const ms = (action.ms as number) ?? 2000;
			execFileSync("/bin/sleep", [String(ms / 1000)]);
			break;
		}
		case "screenshot":
			break;
		default:
			console.error(`[cua-native] unknown action: ${actionType}`);
	}
}

// ── Internal helpers ────────────────────────────────────────────────────

let defaultDisplay = 1;

function getDefaultDisplay(): number {
	return defaultDisplay;
}

export function setDefaultDisplay(idx: number): void {
	defaultDisplay = idx;
}

function getDisplayPointDimensions(): DisplayDimensions {
	const displayIdx = getDefaultDisplay();
	const script = `use framework "AppKit"
set screens to current application's NSScreen's screens()
set idx to ${displayIdx}
if idx > (count of screens) then set idx to 1
set s to item idx of screens
set f to s's frame()
set w to item 1 of item 2 of f
set h to item 2 of item 2 of f
return (w as text) & "x" & (h as text)`;

	try {
		const out = execFileSync("/usr/bin/osascript", ["-e", script], { encoding: "utf-8" }).trim();
		const [wStr, hStr] = out.split("x");
		const w = parseFloat(wStr);
		const h = parseFloat(hStr);
		if (!Number.isNaN(w) && !Number.isNaN(h)) {
			console.error(`[cua-native] display ${displayIdx} point dims: ${w}x${h}`);
			return { width: w, height: h };
		}
	} catch {
		// fall through to default
	}

	console.error("[cua-native] could not query display dims, using 1440x900 fallback");
	return { width: 1440, height: 900 };
}

async function callResponsesApi(
	body: Record<string, unknown>,
): Promise<Record<string, unknown>> {
	const resp = await openaiFetch("/v1/responses", {
		method: "POST",
		body: JSON.stringify(body),
	});

	if (!resp.ok) {
		const errText = await resp.text();
		throw new Error(`CUA API error: ${errText}`);
	}

	const data = (await resp.json()) as Record<string, unknown>;

	if (data.error) {
		throw new Error(`CUA API error: ${JSON.stringify(data.error)}`);
	}

	return data;
}

function findComputerCall(resp: Record<string, unknown>): ComputerCall | null {
	const output = resp.output as Record<string, unknown>[] | undefined;
	if (!Array.isArray(output)) return null;

	for (const item of output) {
		if (item.type === "computer_call") {
			return {
				call_id: (item.call_id as string) ?? "",
				actions: (item.actions as Record<string, unknown>[]) ?? [],
			};
		}
	}
	return null;
}

function extractTextOutput(resp: Record<string, unknown>): string {
	const output = resp.output as Record<string, unknown>[] | undefined;
	if (!Array.isArray(output)) return "Task completed.";

	const texts: string[] = [];
	for (const item of output) {
		if (item.type === "message") {
			const content = item.content as Record<string, unknown>[] | undefined;
			if (Array.isArray(content)) {
				for (const c of content) {
					if (c.type === "output_text" && typeof c.text === "string") {
						texts.push(c.text);
					}
				}
			}
		}
	}

	return texts.length > 0 ? texts.join("\n") : "Task completed.";
}

async function takeBrowserScreenshot(): Promise<string> {
	const result = await browser_command("cua_screenshot", {});
	if (!result.ok) throw new Error("Screenshot failed");

	const b64 = (result.data as Record<string, unknown>)?.base64;
	if (typeof b64 !== "string") throw new Error("No base64 in screenshot response");
	return b64;
}

async function executeCuaAction(action: Record<string, unknown>): Promise<void> {
	const actionType = (action.type as string) ?? "unknown";

	switch (actionType) {
		case "click": {
			const x = Math.round((action.x as number) ?? 0);
			const y = Math.round((action.y as number) ?? 0);
			const button = (action.button as string) ?? "left";
			const keys = (action.keys as string[]) ?? [];
			await browser_command("cua_click", { x, y, button, keys });
			break;
		}
		case "double_click": {
			const x = Math.round((action.x as number) ?? 0);
			const y = Math.round((action.y as number) ?? 0);
			await browser_command("cua_double_click", { x, y });
			break;
		}
		case "type": {
			const text = (action.text as string) ?? "";
			await browser_command("cua_type", { text });
			break;
		}
		case "keypress": {
			const keys = (action.keys as string[]) ?? [];
			await browser_command("cua_keypress", { keys });
			break;
		}
		case "scroll": {
			const x = Math.round((action.x as number) ?? 640);
			const y = Math.round((action.y as number) ?? 450);
			const scroll_x = Math.round((action.scroll_x as number) ?? 0);
			const scroll_y = Math.round((action.scroll_y as number) ?? 0);
			await browser_command("cua_scroll", { x, y, scroll_x, scroll_y });
			break;
		}
		case "drag": {
			const path = action.path ?? [];
			await browser_command("cua_drag", { path });
			break;
		}
		case "move": {
			const x = Math.round((action.x as number) ?? 0);
			const y = Math.round((action.y as number) ?? 0);
			await browser_command("cua_move", { x, y });
			break;
		}
		case "wait": {
			const ms = (action.ms as number) ?? 2000;
			await browser_command("cua_wait", { ms });
			break;
		}
		case "screenshot":
			break;
		default:
			console.error(`[cua] unknown action type: ${actionType}`);
	}
}

// Locate `helpers/native-input.swift` in both dev and packaged-app layouts.
//
// Packaged: `extraResources` in package.json copies the entire `helpers/`
// dir to `Samuel.app/Contents/Resources/helpers/`. We MUST find it there
// rather than at `__dirname/../../helpers/...` (which would resolve inside
// `app.asar`) because the swift binary at `/usr/bin/swift` can't read paths
// inside an asar archive — it'll get ENOENT and CUA fails silently.
//
// Dev: `process.resourcesPath` points at Electron's own resources dir
// (.../node_modules/electron/dist/Electron.app/Contents/Resources/) which
// does NOT contain our helpers, so the existsSync fails and we fall through
// to the cwd/__dirname checks that work in dev.
function findNativeInputHelper(): string {
	const fromResources = join(process.resourcesPath, "helpers", "native-input.swift");
	if (existsSync(fromResources)) return fromResources;

	const fromCwd = join(process.cwd(), "helpers", "native-input.swift");
	if (existsSync(fromCwd)) return fromCwd;

	const fromDirname = resolve(__dirname, "..", "..", "helpers", "native-input.swift");
	if (existsSync(fromDirname)) return fromDirname;

	return "helpers/native-input.swift";
}

function runNativeInput(helper: string, args: string[]): void {
	try {
		execFileSync("/usr/bin/swift", [helper, ...args]);
	} catch (err) {
		const msg = err instanceof Error ? err.message : String(err);
		throw new Error(`native-input failed: ${msg}`);
	}
}

function sleep(ms: number): Promise<void> {
	return new Promise((resolve) => setTimeout(resolve, ms));
}
